### ISSUE TITLE
Tiered packaging

### ISSUE THREAD

**Issue by SimonBiggs** [OpenSaMD/OpenSaMD#132]
Hi @sjswerdloff,

I am considering making a SOUP section within the repository where I can vendor in "software of unknown provenance" and then make tweaks/adjustments as needed. But I'd like to keep that SOUP under the SOUP classification, as in, have it as an external dependency which for the purposes of the medical device is considered external. And therefore, the medical device software needs to implement testing for that SOUP, but the SOUP itself doesn't need to undergo line-by-line review.

My thought is, I would split up the repo into `soup` and `device` directories making it clear what is being considered Device code and what is being considered SOUP code for code review and regulatory process.

In practice this would mean, I wouldn't be asking for your review on SOUP code. Instead the review would need to occur at the interface point where we utilise via the public interface and write and undergo the testing of that utilisation within the Device section of repository. 

Keen for your thoughts on this.

Cheers 🙂,
Simon

**Comment by sjswerdloff** [comment]
If it's SOUP, I think you need to be very careful about "monkey patching".
Any interface between the "monkey patch" and the external code becomes part of what you need to test.
This also means that whatever process one has for evaluating SOUP has to be repeated with any change at all to the SOUP.
That includes addressing the potential for interaction between the items in the SOUP ("all stones can affect the broth and the broth can interact with all of the stones").  So I recommend pinning SOUP stringently in between conscious choices to unpin/update.  If one can automate extensive testing of the SOUP in context, perhaps the CI can take care of the issue.

**Comment by SimonBiggs** [comment]
Yup, most certainly. As a heads up, I most definitely would not be [monkey patching](https://en.wikipedia.org/wiki/Monkey_patch) (run time code patching).

And in fact, the whole reason for bringing in external SOUP into the repository (as opposed to forking it in a separate repository) is actually so that I can have it come under the same CI banner and be able to have the testing automation cover it within the repository.

If upstream developers are willing to accept the changes I am making then I most certainly will err on the side of having any changes be made be able to be contributed upstream. But it is often helpful for upstream developers to see the benefit in a downstream project before having it be included within the upstream library.
